conformance/conformance_go.go

```diff
@@ -0,0 +1,159 @@
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package main
+
+import (
+	"bufio"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"io"
+	"log"
+	"os"
+
+	pb "github.com/google/protobuf/conformance/conformance"
+)
+
+// doTest returns a ConformanceResponse with the results of a test.
+// Tests parse a Protobuf or JSON payload and then output a serialized Protobuf
+// or JSON payload.
+func doTest(req *pb.ConformanceRequest) (*pb.ConformanceResponse, error) {
+	resp := &pb.ConformanceResponse{}
+	test := &pb.TestAllTypes{}
+
+	switch p := req.Payload.(type) {
+	case *pb.ConformanceRequest_ProtobufPayload:
+		if err := proto.Unmarshal(p.ProtobufPayload, test); err != nil {
+			resp.Result = &pb.ConformanceResponse_ParseError{ParseError: err.Error()}
+			return resp, nil
+		}
+	case *pb.ConformanceRequest_JsonPayload:
+		if err := jsonpb.UnmarshalString(p.JsonPayload, test); err != nil {
+			resp.Result = &pb.ConformanceResponse_ParseError{ParseError: err.Error()}
+			return resp, nil
+		}
+	case nil:
+		return resp, errors.New("request didn't have payload")
+	default:
+		return resp, errors.New("request had unknown payload type")
+	}
+
+	switch req.RequestedOutputFormat {
+	case pb.WireFormat_UNSPECIFIED:
+		return resp, errors.New("unspecified output format")
+	case pb.WireFormat_PROTOBUF:
+		b, err := proto.Marshal(test)
+		if err != nil {
+			return resp, err
+		}
+		resp.Result = &pb.ConformanceResponse_ProtobufPayload{ProtobufPayload: b}
+	case pb.WireFormat_JSON:
+		marshal := &jsonpb.Marshaler{}
+		s, err := marshal.MarshalToString(test)
+		if err != nil {
+			return resp, err
+		}
+		resp.Result = &pb.ConformanceResponse_JsonPayload{JsonPayload: s}
+	default:
+		return resp, errors.New("request had unknown output format")
+	}
+
+	return resp, nil
+}
+
+// doTestIO returns true if test ran successfully, falso on legitimate EOF.
+// If EOF is encountered in an unexpected place, returns an error.
+func doTestIO() (bool, error) {
+	r := bufio.NewReader(os.Stdin)
+
+	// Get the length of the next message, which is encoded in the first 4 bytes
+	// as an unsigned integer. If there aren't 4 bytes here, the test is
+	// finished.
+	b, err := r.Peek(4)
+	if len(b) == 0 && err == io.EOF {
+		return false, nil
```

Comment:
Because in this case it is a success. If we got 0 bytes it means that the test is finished. (If we got 1, 2, or 3 bytes it means something is wrong / unexpected)

Should I add a comment? This is the same behavior as the Python and Ruby conformance tests.